samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -263,6 +304,18 @@ public String getNextOpId(OpCode opCode) {
     return Collections.unmodifiableSet(broadcastStreams);
   }
 
+  public Map<String, InputDescriptor> getInputDescriptors() {
+    return Collections.unmodifiableMap(inputDescriptors);
+  }
+
+  public Map<String, OutputDescriptor> getOutputDescriptors() {
+    return Collections.unmodifiableMap(outputDescriptors);
+  }
+
+  public Set<SystemDescriptor> getSystemDescriptors() {
```

Comment:
Does SystemDescriptor implement hashCode/equals?

Reply:
No, but reference equality should be sufficient here. The constraint of 1 instance per system name is enforced when getting input/output/intermediate streams and setting default system descriptor.

Reply:
Just seems to go a bit against convention. It might be good to put a comment about this here to help avoid someone assuming that there is some non-reference deduping going on.